File: cis_interface/examples/timed_pipe/src/timed_pipe_src.py
from __future__ import print_function
import sys
from cis_interface.interface.CisInterface import CisOutput


def run(args):
    msg_count = int(args[0])
    msg_size = int(args[1])
    print('Hello from Python pipe_src: msg_count = %d, msg_size = %d' % (
        msg_count, msg_size))

    # Ins/outs matching with the the model yaml
    outq = CisOutput('output_pipe')
    print("pipe_src(P): Created I/O channels")

    # Send test message multiple times
    test_msg = '0' * msg_size
    count = 0
    for i in range(msg_count):
        ret = outq.send(test_msg)
        if not ret:
            print('pipe_src(P): SEND ERROR ON MSG %d' % i)
            break
        count += 1
    ret = outq.send_eof()
    if not ret:
        print('pipe_src(P): SEND ERROR ON EOF')

    print('Goodbye from Python source. Sent %d messages.' % count)

    return 0
    

if __name__ == '__main__':
    run(sys.argv[1:])

File: yggdrasil/examples/timed_pipe/src/timed_pipe_src.f90
program main
  use fygg

  character(len=32) :: arg
  integer :: msg_count, msg_size
  type(yggcomm) :: outq
  logical :: ret
  integer :: i, count
  integer :: exit_code = 0
  character(len=:), allocatable :: test_msg

  if (command_argument_count().ne.2) then
     write(*, '("Error in Fortran pipe_src: The message count and size &
          &must be provided as input arguments.")')
     stop 1
  end if

  call get_command_argument(1, arg)
  read(arg, *) msg_count
  call get_command_argument(2, arg)
  read(arg, *) msg_size
  write(*, '("Hello from Fortran pipe_src: msg_count = ",i5.1,", &
       &msg_size = ",i5.1)') msg_count, msg_size

  ! Ins/outs matching with the the model yaml
  outq = ygg_output("output_pipe")
  write(*, '("pipe_src(F): Created I/O channels")')

  ! Create test message
  allocate(character(len=(msg_size + 1)) :: test_msg)
  do i = 1, msg_size
     test_msg(i:i) = '0'
  end do
  test_msg((msg_size+1):(msg_size+1)) = c_null_char

  ! Send test message multiple times
  count = 0
  do i = 1, msg_count
     ret = ygg_send(outq, test_msg, msg_size)
     if (.not.ret) then
        write(*, '("pipe_src(F): SEND ERROR ON MSG ",i5.1)') i
        exit_code = -1
        exit
     end if
     count = count + 1
  end do
  
  write(*, '("Goodbye from Fortran source. Sent ",i5.1," messages.")') count
  if (allocated(test_msg)) deallocate(test_msg)
  if (exit_code.lt.0) then
     stop 1
  end if

end program main
